src/TTImp/Unelab.idr

```diff
@@ -63,98 +63,111 @@ mutual
   ||| name x.
   |||
   ||| We will try to do our best...
-  unelabCase : {auto c : Ref Ctxt Defs} ->
+  unelabCase : {vars : _} ->
+               {auto c : Ref Ctxt Defs} ->
                List (Name, Nat) ->
-               Name -> List IArg -> IRawImp ->
-               Core IRawImp
-  unelabCase nest n args orig
+               Env Term vars ->
+               Name ->
+               List (Term vars) ->
+               Core (Maybe IRawImp)
+  unelabCase nest env n args
       = do defs <- get Ctxt
            Just glob <- lookupCtxtExact n (gamma defs)
-                | Nothing => pure orig
+                | Nothing => pure Nothing
            let PMDef _ pargs treect _ pats = definition glob
-                | _ => pure orig
+                | _ => pure Nothing
            let Just argpos = findArgPos treect
-                | _ => pure orig
+                | _ => pure Nothing
            if length args == length pargs
-              then mkCase pats argpos 0 args
-              else pure orig
+              then mkCase pats argpos args
+              else pure Nothing
     where
       -- Need to find the position of the scrutinee to rebuild original
       -- case correctly
       findArgPos : CaseTree as -> Maybe Nat
       findArgPos (Case idx p _ _) = Just idx
       findArgPos _ = Nothing
 
-      idxOrDefault : Nat -> a -> List a -> a
-      idxOrDefault Z def (x :: _) = x
-      idxOrDefault (S k) def (_ :: xs) = idxOrDefault k def xs
-      idxOrDefault _ def [] = def
+      idxOrMaybe : Nat -> List a -> Maybe a
+      idxOrMaybe Z (x :: _) = Just x
+      idxOrMaybe (S k) (_ :: xs) = idxOrMaybe k xs
+      idxOrMaybe _ [] = Nothing
 
-      getNth : Nat -> Term vars -> Term vars
-      getNth n tm
-          = case getFnArgs tm of
-                 (f, as) => idxOrDefault n f as
+      -- TODO: some utility like this should probably be implemented in Core
+      substVar : List (Var vs) -> Term vs -> Term vs -> Term vs
```

Comment:
Can't we get
```
substVar : List (Var vs, Term vs) -> Term vs -> Term vs
```
directly rather than using `foldl` on the single substVar with the repeated traversals this implies?

Reply:
done